### PR TITLE
TCP multi host/port support

### PR DESCRIPTION
**Pull request by nshamshoum**

### CODE REVIEW

File: common.sh [hunk 1]
```diff
@@ -221,8 +260,8 @@ prep_for_tune_and_iperf_test() {
     ssh "${SERVER_TRUSTED}" iperf3 -v
     ssh "${CLIENT_TRUSTED}" cat /proc/cmdline
     ssh "${SERVER_TRUSTED}" cat /proc/cmdline
-    #ssh "${CLIENT_TRUSTED}" iperf -v
-    #ssh "${SERVER_TRUSTED}" iperf -v
+    ssh "${CLIENT_TRUSTED}" iperf -v
+    ssh "${SERVER_TRUSTED}" iperf -v
```
Comment (blochl) [inline common.sh:264]: Why do you bring back iperf2 version check? It's not used anymore...

File: common.sh [hunk 2]
```diff
@@ -289,7 +328,7 @@ run_iperf3() {
     ssh "${SERVER_TRUSTED}" "bash -s" -- < "${scriptdir}/run_iperf3_servers.sh" \
         "${PROC}" "${SERVER_NUMA_NODE}" "${SERVER_LOGICAL_NUMA_PER_SOCKET}" \
         "${SERVER_BASE_NUMA}" "${BASE_TCP_PORT}" &
-    if [ "${DUPLEX}" = "FULL" ]; then
+    if [ $DUPLEX ]; then
```
Comment (blochl) [inline common.sh:331]: Why is this?

File: common.sh [hunk 3]
```diff
@@ -301,7 +340,7 @@ run_iperf3() {
         "${RESULT_FILE}" "${PROC}" "${CLIENT_NUMA_NODE}" "${CLIENT_LOGICAL_NUMA_PER_SOCKET}" \
         "${CLIENT_BASE_NUMA}" "${SERVER_IP[$((P%IP_AMOUNT))]}" \
         "${BASE_TCP_PORT}" "${THREADS}" "${TIME}" &
-    if [ "${DUPLEX}" = "FULL" ]; then
+    if [ $DUPLEX ]; then
```
Comment (blochl) [inline common.sh:343]: Why is this?

File: ngc_tcp_test.sh [hunk 4]
```diff
@@ -1,13 +1,14 @@
 #!/bin/bash
-# NGC Certification TCP test v2.3
-# Owner: amira@nvidia.com
-#
```
Comment (blochl) [inline ngc_tcp_test.sh:4]: This is not related to this commit.

File: common.sh [hunk 5]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
     done
     exit 1
 }
+
+get_numa_nodes_array() {
```
Comment (blochl) [inline common.sh:434]: Please verify the indentations in this function.

File: common.sh [hunk 6]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+                ITER=$((ITER+1))
+        done
+        echo "${DEVCIES_NUMA[*]}"
+}
+
+#global params :
+#NUMA_NODES
+get_available_cores_per_device(){
+
+    local SERVER_TRUSTED=$1
```
Comment (blochl) [inline common.sh:459]: No need for an empty line.

File: common.sh [hunk 7]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+enable_aRFS() {
+    local SERVER=$1
+    local SERVER_NETDEV=$2
+    #TODO: check if supported
+    ssh "${SERVER}" "sudo ethtool -K ${SERVER_NETDEV} ntuple off"
+    ssh "${SERVER}" "sudo bash -c 'echo 0 > /proc/sys/net/core/rps_sock_flow_entries'"
+    ## 32768 1170
+    ssh "${SERVER}" "for f in /sys/class/net/${SERVER_NETDEV}/queues/rx-*/rps_flow_cnt; do sudo bash -c \"echo '0' > \${f}\"; done"
+}
+enable_flow_stearing(){
```
Comment (blochl) [inline common.sh:631]: Probably better to separate functions with an empty line.

File: common.sh [hunk 8]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+    log "INFO: done attempting to delete any existing rules, ethtool -U $SERVER_NETDEV delete "
+    sleep 1
+    local i=0
+    for ((; i < NUM_INST; i++))
+    do
+        ssh "${SERVER}" "sudo ethtool -U $SERVER_NETDEV flow-type tcp4 dst-port $((10000*(index+1) + i)) loc $i queue $i" &>> /dev/null
+        ssh "${SERVER}" "sudo ethtool -U $SERVER_NETDEV flow-type tcp4 src-port $((10000*(index+1) + i)) loc $((i+NUM_INST+1)) queue $((i+NUM_INST+1))" &>> /dev/null
+    done
+}
+is_SPR() {
```
Comment (blochl) [inline common.sh:649]: Probably better to separate functions with an empty line.

File: common.sh [hunk 9]
```diff
@@ -119,6 +126,38 @@ get_n_min_distances() {
...
+        CLIENT_IPS+=("$(ssh "${CLIENT_TRUSTED}" "ip a sh ${CLIENT_NETDEVS[${#CLIENT_NETDEVS[@]}-1]}" | grep -ioP '(?<=inet )\d+\.\d+\.\d+\.\d+' | xargs | tr ' ' ',')")
+        [ -z "${CLIENT_IPS[${#CLIENT_IPS[@]}-1]}" ] &&
+            fatal "Can't find a client IP associated with the net device '${CLIENT_NETDEVS[${#CLIENT_NETDEVS[@]}-1]}'." ||
+            log "INFO: Found $(awk -F',' '{print NF}' <<<"${CLIENT_IPS[${#CLIENT_IPS[@]}-1]}") IPs associated with the client net device '${CLIENT_NETDEVS[${#CLIENT_NETDEVS[@]}-1]}'."
+        [ -n "${CLIENT_NETDEVS}" ] ||
+        fatal "Can't find client net device. Did you mean to specify IB device as '${CLIENT_DEVICE}'?"
+    done
+}
+
 get_netdev_from_ibdev() {
```
Comment (blochl) [inline common.sh:161]: Please use this function for netdev detection, to be able to run between DPUs.

File: common.sh [hunk 10]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+
+    #Copy the file to the servers to ensure running all iperf3 clients at the same time
+    scp ${iperf_clients_to_run_client_side} ${CLIENT_TRUSTED}:${iperf_clients_to_run_client_side}
+    if [ "$DUPLEX"  = true ]
+    then
+        scp ${iperf_clients_to_run_server_side} ${SERVER_TRUSTED}:${iperf_clients_to_run_server_side}
+        ssh ${SERVER_TRUSTED} "sleep 0.01 ; bash ${iperf_clients_to_run_server_side}" &>> /dev/null &
+    fi
+    #Run all iperf clients
+    log "INFO:iperf3 clietns start to run , wait for ${TEST_DURATION} for the test to finish"
```
Comment (blochl) [inline common.sh:812]: Please add time units (... wait for ${TEST_DURATION} **s** for ...).

File: ngc_tcp_test.sh [hunk 11]
```diff
@@ -39,103 +52,78 @@ do
...
+FORCE_EXIT=false
+#Set number of cores to use and apply tuning according to #core and list of cores
+#Expected global params :
+#CLIENT_TRUSTED,CLIENT_DEVICES,SERVER_TRUSTED,SERVER_DEVICES,NUM_INST,opt_proc,CORES_ARRAY
+tune_tcp
+#Relaxed ordring was disabled - user need to restart the driver so that the change take affect.
+[ $FORCE_EXIT = true ] && echo -e "${RED}Please restart driver after disabling relaxed ordering (RO), and run the script again ${NC}" && exit 1
 
-# Toggle interfaces down/up so channels allocation will be according to actual IRQ affinity
-ssh "${SERVER_TRUSTED}" "sudo ip l set ${SERVER_NETDEV} down; sudo ip l set ${SERVER_NETDEV} up; sudo ip a add ${SERVER_IP[0]}/24 broadcast + dev ${SERVER_NETDEV}"
```
Comment (blochl) [inline ngc_tcp_test.sh:131]: Are you sure that IRQ affinity is set correctly if you disabled the down/up cycle?

File: common.sh [hunk 12]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+    local dev_idx=0
+    for ((; dev_idx<NUM_DEVS; dev_idx++))
+    do
+        local OFFSET_S=$((dev_idx*NUM_CORES_PER_DEVICE ))
+        for i in `seq 0 $((NUM_INST-1))`; do
+            sleep 0.1
+            index=$((i+OFFSET_S))
+            core="${CORES_ARRAY[index]}"
+            prt=$((BASE_TCP_POTR + 10000*dev_idx + i ))
+            ssh "${SERVER_TRUSTED}" "taskset -c $core iperf3 -s -p $prt --one-off &>> /dev/null " &
```
Comment (blochl) [inline common.sh:739]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 13]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+            log "INFO: run iperf3 server on ${SERVER_TRUSTED} -taskset -c $core iperf3 -s -p $prt --one-off &>> /dev/null "
```
Comment (blochl) [inline common.sh:740]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 14]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+        if [ "$DUPLEX"  = true ]
+        then
+            local OFFSET_C=$((dev_idx*NUM_CORES_PER_DEVICE+ NUM_DEVS*NUM_CORES_PER_DEVICE))
+            for i in `seq 0 $((NUM_INST-1))`
+            do
+                sleep 0.1
+                index=$(( i+OFFSET_C ))
+                core="${CORES_ARRAY[index]}"
+                prt=$((BASE_TCP_POTR + 10000*dev_idx + i ))
+                ssh "${CLIENT_TRUSTED}" "taskset -c $core iperf3 -s -p $prt --one-off &>> /dev/null " &
```
Comment (blochl) [inline common.sh:752]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 15]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+                log "INFO: run iperf3 server on ${CLIENT_TRUSTED} core index= $index - taskset -c $core iperf3 -s -p $prt --one-off &>> /dev/null "
```
Comment (blochl) [inline common.sh:753]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 16]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+            done
+        fi
+    done
+
+    #Copy the file to the servers to ensure running all iperf3 clients at the same time
+    scp ${iperf_clients_to_run_client_side} ${CLIENT_TRUSTED}:${iperf_clients_to_run_client_side}
+    if [ "$DUPLEX"  = true ]
+    then
+        scp ${iperf_clients_to_run_server_side} ${SERVER_TRUSTED}:${iperf_clients_to_run_server_side}
+        ssh ${SERVER_TRUSTED} "sleep 0.01 ; bash ${iperf_clients_to_run_server_side}" &>> /dev/null &
```
Comment (blochl) [inline common.sh:809]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 17]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+    fi
+    #Run all iperf clients
+    log "INFO:iperf3 clietns start to run , wait for ${TEST_DURATION} for the test to finish"
+    ssh ${CLIENT_TRUSTED} "bash ${iperf_clients_to_run_client_side}" &>> /dev/null
```
Comment (blochl) [inline common.sh:813]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 18]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+    ssh "${SERVER}" "for f in /sys/class/net/${SERVER_NETDEV}/queues/rx-*/rps_flow_cnt; do sudo bash -c \"echo '0' > \${f}\"; done"
+}
+enable_flow_stearing(){
+    local SERVER=$1
+    local SERVER_NETDEV=$2
+    local index=$3
+    local j=0
+    for ((; j<30; j++))
+    do
+        ssh "${SERVER}" "sudo ethtool -U $SERVER_NETDEV delete $j &>> /dev/null"
```
Comment (blochl) [inline common.sh:638]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 19]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+    done
+    log "INFO: done attempting to delete any existing rules, ethtool -U $SERVER_NETDEV delete "
+    sleep 1
+    local i=0
+    for ((; i < NUM_INST; i++))
+    do
+        ssh "${SERVER}" "sudo ethtool -U $SERVER_NETDEV flow-type tcp4 dst-port $((10000*(index+1) + i)) loc $i queue $i" &>> /dev/null
+        ssh "${SERVER}" "sudo ethtool -U $SERVER_NETDEV flow-type tcp4 src-port $((10000*(index+1) + i)) loc $((i+NUM_INST+1)) queue $((i+NUM_INST+1))" &>> /dev/null
```
Comment (blochl) [inline common.sh:646]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh [hunk 20]
```diff
@@ -391,3 +430,458 @@ get_cudas_per_rdma_device() {
...
+            s_core=0
+            s_add_dumy_core=0
+        fi
+        if (( c_core < 0))
+        then
+            c_core=0
+            c_add_dumy_core=0
+        fi
+        ssh "${SERVER_TRUSTED}" sudo set_irq_affinity_cpulist.sh "$(tr " " "," <<< "${CORES_ARRAY[@]:s_core:$((NUM_INST+s_add_dumy_core))}")" "${SERVER_DEVICES[i]}" &>> /dev/null
+        ssh "${CLIENT_TRUSTED}" sudo set_irq_affinity_cpulist.sh "$(tr " " "," <<< "${CORES_ARRAY[@]:c_core:$((NUM_INST+c_add_dumy_core))}")" "${CLIENT_DEVICES[i]}" &>> /dev/null
```
Comment (blochl) [inline common.sh:713]: Why append (>>) to `/dev/null` and not redirect (>)?

File: common.sh (hunk 2 above)
Comment (nshamshoum) [reply common.sh:331]: I have changed this to true/false instead of FULL/HALF respectively, added = "true" to be more clear

File: common.sh (hunk 3 above)
Comment (nshamshoum) [reply common.sh:343]: same as before

File: common.sh (hunk 5 above)
Comment (nshamshoum) [reply common.sh:434]: The indentation is correct since the indented section is nested under if

File: ngc_tcp_test.sh (hunk 11 above)
Comment (nshamshoum) [reply ngc_tcp_test.sh:131]: Yes - as long as there is no traffic it should pick-up the new affinity - I yet i have added you code to be in the safe dise

File: common.sh [hunk 21]
```diff
@@ -289,7 +328,7 @@ run_iperf3() {
     ssh "${SERVER_TRUSTED}" "bash -s" -- < "${scriptdir}/run_iperf3_servers.sh" \
         "${PROC}" "${SERVER_NUMA_NODE}" "${SERVER_LOGICAL_NUMA_PER_SOCKET}" \
         "${SERVER_BASE_NUMA}" "${BASE_TCP_PORT}" &
-    if [ "${DUPLEX}" = "FULL" ]; then
+    if [ "$DUPLEX" = true ]; then
```
Comment (blochl) [inline common.sh:331]: Here no quotes but quotes in line 343. Let's be consistent. I like quotes, but I don't mind that both will be without quotes, as long as it's consistent.

Comment (nshamshoum) [reply common.sh]: Thanks, i have add quotes

**Review by blochl — APPROVED** [review]
Awesome job, @nshamshoum! Thanks!!
